src/Common/src/CoreLib/System/IO/PathInternal.cs

```diff
@@ -117,24 +126,51 @@ internal static bool AreRootsEqual(string first, string second, StringComparison
         internal static string RemoveRelativeSegments(string path, int rootLength)
         {
             Debug.Assert(rootLength > 0);
-            bool flippedSeparator = false;
 
+            Span<char> initialBuffer = stackalloc char[MaxStackPath];
+            ValueStringBuilder sb = new ValueStringBuilder(initialBuffer);
+
+            bool pathNotChanged = RemoveRelativeSegmentsHelper(path.AsSpan(), rootLength, ref sb);
+
+            if (sb.Length >= rootLength && !pathNotChanged)
+                return sb.ToString();
+
+            // If we haven't changed the source path, return the original
+            return pathNotChanged ? path : path.Substring(0, rootLength);
+        }
+
+        /// <summary>
+        /// Try to remove relative segments from the given path (without combining with a root).
+        /// </summary>
+        /// <param name="rootLength">The length of the root of the given path</param>
+        internal static void RemoveRelativeSegments(ReadOnlySpan<char> path, int rootLength, ref ValueStringBuilder sb)
+        {
+            Debug.Assert(rootLength > 0);
+            RemoveRelativeSegmentsHelper(path, rootLength, ref sb);
+
+            if (sb.Length < rootLength)
+            {
+                sb.Length = rootLength;
+                sb[sb.Length - 1] = path[rootLength - 1];
```

Comment:
What are you doing here?

Reply:
https://github.com/dotnet/corefx/pull/28835/files#diff-200474c05bdb868e1d141aeeb98cbf18R156